src/normalize/ioctl.rs

```diff
@@ -0,0 +1,301 @@
+    }
+    if vma_flags & PROCMAP_QUERY_VMA_EXECUTABLE != 0 {
+        mode |= 0b0010;
+    }
+    mode
+}
+
+
+/// The caller is responsible for checking that the returned `MapsEntry`
+/// actually covers the provided address. If it does not, it represents
+/// the next known entry.
+pub(crate) fn query_procmap(
+    fd: BorrowedFd,
+    pid: Pid,
+    addr: Addr,
+    build_id: bool,
+) -> Result<Option<MapsEntry>> {
+    let mut path_buf = MaybeUninit::<[u8; 4096]>::uninit();
+    let mut build_id_buf = MaybeUninit::<[u8; 56]>::uninit();
+    let mut query = procmap_query {
+        size: size_of::<procmap_query>() as _,
+        query_flags: (PROCMAP_QUERY_COVERING_OR_NEXT_VMA
+            // NB: Keep filter flags in sync with `filter_relevant`
+            //     function.
+            | PROCMAP_QUERY_VMA_READABLE
+            | PROCMAP_QUERY_VMA_EXECUTABLE) as _,
+        query_addr: addr,
+        vma_name_addr: path_buf.as_mut_ptr() as _,
+        vma_name_size: size_of_val(&path_buf) as _,
+        build_id_addr: if build_id {
+            build_id_buf.as_mut_ptr() as _
+        } else {
+            0
+        },
+        build_id_size: if build_id {
+            size_of_val(&build_id_buf) as _
+        } else {
+            0
+        },
+        ..Default::default()
+    };
+
+    // SAFETY: The `procmap_query` pointer is valid because it comes
+    //         from a reference.
+    let rc = unsafe {
+        ioctl(
+            fd.as_raw_fd(),
+            PROCMAP_QUERY as _,
+            &mut query as *mut procmap_query,
+        )
+    };
+    if rc < 0 {
+        let err = io::Error::last_os_error();
+        match err.raw_os_error() {
+            Some(e) if e == ENOTTY => {
+                return Err(Error::with_unsupported("PROCMAP_QUERY is not supported"))
+            }
+            Some(e) if e == ENOENT => return Ok(None),
+            _ => (),
+        }
+        return Err(Error::from(err))
+    }
+
+    // SAFETY: The kernel will have set the member to a valid value.
+    let path_buf = unsafe { path_buf.assume_init_ref() };
+
+    let maps_file = PathBuf::from(format!(
+        "/proc/{pid}/map_files/{:x}-{:x}",
+        query.vma_start, query.vma_end
+    ));
+    let symbolic_path =
+        bytes_to_path(&path_buf[0..query.vma_name_size.saturating_sub(1) as usize])?.to_path_buf();
+    let mut entry = MapsEntry {
+        range: query.vma_start..query.vma_end,
+        mode: vma_flags_to_mode(query.vma_flags),
+        offset: query.vma_offset,
+        path_name: Some(PathName::Path(EntryPath {
+            maps_file,
+            symbolic_path,
+            _non_exhaustive: (),
+        })),
+        build_id: None,
+    };
+
+    if build_id {
+        // SAFETY: The kernel will have set the member to a valid value
+        //         because we asked it to.
+        let build_id_buf = unsafe { build_id_buf.assume_init_ref() };
+        let build_id = build_id_buf[0..query.build_id_size as usize].to_vec();
```

Comment:
a bit wasteful to allocate an empty vector if there is no build ID? maybe have one static empty array to represent "no build ID" condition that would be shared by all VMAs with no build ID?

Reply:
I think we should probably represent it as `None` if no build ID is being reported. Will change.